### PR TITLE
Fix to-py conversion with duplicate enum values

### LINKED ISSUE #5400: [BUG]  cpdef Enum does not support enums with the same value
### Describe the bug

When a cpdef enum is used and has duplicate values, Cython fails to compile the *.pyx file.

For example:

```
cdef enum SomeType:
    BYTE = 1
    UINT8 = 1
    UINT16 = 2
    UINT32 = 3
```

I belive this is because a switch is being used during optimization (maybe here? https://github.com/cython/cython/blob/master/Cython/Utility/CpdefEnums.pyx#L154)

Perhaps a bug with enums here: https://github.com/cython/cython/blob/master/Cython/Compiler/Optimize.py#L1246

### Code to reproduce the behaviour:

```diff
diff --git a/tests/run/switch.pyx b/tests/run/switch.pyx
index dfd734f..a88e679 100644
--- a/tests/run/switch.pyx
+++ b/tests/run/switch.pyx
@@ -333,7 +333,7 @@ def compile_time_tuple_constant(int x):
 
 cdef enum X:
     a = 1
-    b
+    b = 1
     c
     d
     e = 10
```

```
$  python3 runtests.py -vv switch

Compiler output:
switch.cpp:5165:10: error: duplicate case value: '__pyx_e_6switch_a' and '__pyx_e_6switch_b' both equal '1'
    case __pyx_e_6switch_b:
         ^
switch.cpp:5163:10: note: previous case defined here
    case __pyx_e_6switch_a:
         ^
switch.cpp:5356:10: error: duplicate case value: '__pyx_e_6switch_a' and '__pyx_e_6switch_b' both equal '1'
    case __pyx_e_6switch_b:
         ^
switch.cpp:5354:10: note: previous case defined here
    case __pyx_e_6switch_a:
         ^
switch.cpp:5601:10: error: duplicate case value: '__pyx_e_6switch_a' and '__pyx_e_6switch_b' both equal '1'
    case __pyx_e_6switch_b:
         ^
switch.cpp:5599:10: note: previous case defined here
    case __pyx_e_6switch_a:
         ^
3 errors generated.

```

### Expected behaviour

Should be valid and compile. Nothing prevents C or C++ enums from having duplicate values. This breaks existing code during upgrade.

### OS

Linux and macos

### Python version

tested 3.8 and 3.11

### Cython version

3.0.0b2

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by da-woods**
1. Where possible, track the equivalent integer value for an enum type as they're defined.
2. Use this to determine if the switch optimization can be used (if present)
3. If not present, disable the switch optimization for enum to py-enum conversion, since it's duplicates are pretty likely in this case.

Fixes #5400

### CODE REVIEW

File: Cython/Compiler/Nodes.py [hunk 1]
```diff
@@ -1718,8 +1718,14 @@ def analyse_declarations(self, env):
         if self.items is not None:
             if self.in_pxd and not env.in_cinclude:
                 self.entry.defined_in_pxd = 1
+            # For extern enums we can't reason about their equivalent int values because
+            # we can't see them.
+            last_enum_value = 0 if self.visibility != 'extern' else None
             for item in self.items:
-                item.analyse_declarations(scope, self.entry)
+                value_entry = item.analyse_declarations(scope, self.entry, last_enum_value)
+                last_enum_value = value_entry.equivalent_enum_value
```
Comment (da-woods) [inline Cython/Compiler/Nodes.py:1726]:
```suggestion
                last_enum_value = value_entry.equivalent_enum_value if self.visibility != 'extern' else None
```

File: Cython/Compiler/Nodes.py [hunk 2]
```diff
@@ -1752,7 +1758,7 @@ class CEnumDefItemNode(StatNode):
 
     child_attrs = ["value"]
 
-    def analyse_declarations(self, env, enum_entry):
+    def analyse_declarations(self, env, enum_entry, incremental_enum_value):
```
Comment (scoder) [inline Cython/Compiler/Nodes.py:1761]: This should better not be called `analyse_declarations()`. It has a different signature and returns a value. Actually, it analyses the declarations and changes the node state accordingly, but then still returns a value. That's not great. It should do either or, not both.

File: Cython/Compiler/Nodes.py [hunk 3]
```diff
@@ -1718,8 +1718,14 @@ def analyse_declarations(self, env):
         if self.items is not None:
             if self.in_pxd and not env.in_cinclude:
                 self.entry.defined_in_pxd = 1
+            # For extern enums we can't reason about their equivalent int values because
+            # we can't see them.
+            last_enum_value = 0 if self.visibility != 'extern' else None
             for item in self.items:
-                item.analyse_declarations(scope, self.entry)
+                value_entry = item.analyse_declarations(scope, self.entry, last_enum_value)
+                last_enum_value = value_entry.equivalent_enum_value if self.visibility != 'extern' else None
```
Comment (scoder) [inline Cython/Compiler/Nodes.py:1726]: Can't `equivalent_enum_value` simply be `None` if `visibility == 'extern'`? That would avoid the additional condition.

File: Cython/Compiler/Nodes.py [hunk 4]
```diff
@@ -1769,10 +1775,27 @@ def analyse_declarations(self, env, enum_entry):
             self.value, self.pos, cname=cname,
             visibility=enum_entry.visibility, api=enum_entry.api,
             create_wrapper=enum_entry.create_wrapper and enum_entry.name is None)
+
+        enum_value = None
+        if self.value:
+            if self.value.is_literal:
+                enum_value = int(self.value.value)
+            else:
+                value_entry = getattr(self.value, 'entry', None)
```
Comment (scoder) [inline Cython/Compiler/Nodes.py:1784]: It would feel better to spell out the cases where `.entry` exists, rather than having it available *sometimes*. It's unclear from the code what cases those are.

File: Cython/Compiler/PyrexTypes.py [hunk 5]
```diff
@@ -4448,7 +4469,8 @@ def create_to_py_utility_code(self, env):
                     "module_name": module_name,
                     "is_flag": True,
                     },
-            outer_module_scope=self.entry.scope  # ensure that "name" is findable
+            outer_module_scope=self.entry.scope,  # ensure that "name" is findable
+            compiler_directives = directives,
```
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/PyrexTypes.py:4473]: > Does this need to be done here, or can the SwitchTransform do it?

I wondered about this. It'd essentially mean that the switch transform wouldn't be applied to (most?) extern enums. Right now we apply it but document it as a possible cause of invalid C code. So it would be removing an optimization in some cases.

I was trying to do the thing with the smallest behaviour change, but either option is ok

File: Cython/Compiler/Nodes.py (hunk 2 above)
Comment (da-woods) [reply Cython/Compiler/Nodes.py:1761]: The return value is easy to get rid of by assigning entry to the node so I'll definitely do that.

I'd argue that it already has a non-standard signature though, so I'm not making it too much worse by adding an argument

File: Cython/Compiler/PyrexTypes.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dalcinl) [reply Cython/Compiler/PyrexTypes.py:4473]: Cython assuming extern enums can be used with the switch statement had been a problem for me, and I'm currently using `#cython: optimize.use_switch=False` in my projects.

File: Cython/Compiler/Nodes.py (hunk 3 above)
Comment (da-woods) [reply Cython/Compiler/Nodes.py:1726]: It looks like you can specify a value for elements in extern enums:

```
cdef extern from *:
    enum E:
        A = 1
        B
```

In this case I think it's reasonable for Cython to trust what the user says about `A` and use it as `equivalent_enum_value`, but not to infer anything about `B` from this.

Your proposal would work though - it's not using what we've been told about `A`

File: Cython/Compiler/PyrexTypes.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (da-woods) [reply Cython/Compiler/PyrexTypes.py]: For now I've deduplicated the code but not changed the behaviour of `SwitchTransform`.

I think the way to do it would be to make `optimize.use_switch=False` still turn on `SwitchTransform`, but only in  cases where we absolutely know it's unconditionally safe (i.e. no unknown values). Then in the enum utility code, we'd always turn off `optimize.use_switch` and let the transform decide whether it's safe. That feels like a larger change that's probably outside this PR.

File: Cython/Compiler/Nodes.py (hunk 2 above)
Comment (da-woods) [reply Cython/Compiler/Nodes.py:1762]: I've removed the return value. I can rename it as `analyse_enum_item_declarations` (or something like that) if you want, but for now, I'm still claiming it already had a non-standard signature.

File: Cython/Compiler/PyrexTypes.py [hunk 6]
```diff
@@ -4221,7 +4221,49 @@ def cpp_optional_check_for_null_code(self, cname):
...
+            underlying_type_str = "int"
+
+        env.use_utility_code(CythonUtilityCode.load(
+            "EnumTypeToPy", "CpdefEnums.pyx",
+            context={"funcname": self.to_py_function,
+                    "name": self.name,
+                    "items": tuple(self.values),
+                    "underlying_type": underlying_type_str,
+                    "module_name": module_name,
+                    "is_flag": True,
```
Comment (da-woods) [inline Cython/Compiler/PyrexTypes.py:4259]:
```suggestion
                    "is_flag": not self.is_cpp_enum,
```

**Comment by da-woods** [comment]
@scoder Obviously too late for b3, but I think this should be in the next release since it will break some existing code unnecessarily

File: Cython/Compiler/Nodes.py (hunk 2 above)
Comment (scoder) [reply Cython/Compiler/Nodes.py:1762]: > I'm still claiming it already had a non-standard signature.

Then let's rename it. IMHO, that should have been done when it was first introduced. Maybe just `analyse_enum_declarations`? Or `…_enum_item_…`, as you proposed. I don't mind, just something specific.

File: Cython/Compiler/PyrexTypes.py [hunk 7]
```diff
@@ -4221,7 +4221,49 @@ def cpp_optional_check_for_null_code(self, cname):
         return "(%s.has_value())" % cname
 
 
-class CppScopedEnumType(CType):
+class EnumMixin(object):
+    """
+    Common implementation details for C and C++ enums.
+    """
+
+    def create_to_py_utility_code_implementation(self, env):
```
Comment (scoder) [inline Cython/Compiler/PyrexTypes.py:4229]: The semantic difference between `create_to_py_utility_code` and `create_to_py_utility_code_implementation` isn't exactly obvious. I propose making it specific to the `EnumMixin` by renaming it to:
```suggestion
    def create_enum_to_py_utility_code(self, env):
```

File: Cython/Compiler/Nodes.py [hunk 8]
```diff
@@ -1718,8 +1718,17 @@ def analyse_declarations(self, env):
...
+            # we don't know if their definition is complete.
+            is_declared_enum = self.visibility != 'extern'
+
+            next_int_enum_value = 0 if is_declared_enum else None
             for item in self.items:
-                item.analyse_declarations(scope, self.entry)
+                item.analyse_enum_declarations(scope, self.entry, next_int_enum_value)
+                if is_declared_enum:
+                    next_int_enum_value = (
+                        item.entry.enum_int_value or next_int_enum_value) + 1
```
Comment (scoder) [inline Cython/Compiler/Nodes.py:1731]: I now wonder if this is actually safe. What if users declare their enums in an unusual order, like:
```
cdef enum x:
    d = 4
    a = 0
    b, c
```
Wouldn't we lose the `a = 0` in this case? I think we need to test for `None` here instead of a simple truthiness check. (Probably worth adding a declaration like this as a test.)

**Comment by scoder** [comment]
Triggering a fresh CI run.

**Review by scoder — APPROVED** [review]

**Comment by scoder** [comment]
There were a few test failures before the merge (I had invested a couple of days lately to get the master branch build cleanly again). Could you look into fixing those? They cause build failures on the master branch now.
